java/connector-node/risingwave-sink-jdbc/src/main/java/com/risingwave/connector/JDBCSink.java

```diff
@@ -203,48 +221,51 @@ private PreparedStatement prepareDeleteStatement(SinkRow row) {
 
     @Override
     public void write(Iterator<SinkRow> rows) {
+        var stagingStatements = new TreeMap<OpType, PreparedStatement>();
         while (rows.hasNext()) {
             try (SinkRow row = rows.next()) {
-                PreparedStatement stmt;
                 if (row.getOp() == Data.Op.UPDATE_DELETE) {
                     updateFlag = true;
                     continue;
                 }
-
                 if (config.isUpsertSink()) {
-                    stmt = prepareForUpsert(row);
+                    if (row.getOp() == Data.Op.DELETE) {
+                        var stmt = prepareDeleteStatement(row);
+                        stagingStatements.put(OpType.DELETE, stmt);
+                    } else {
+                        var stmt = prepareUpsertStatement(row);
+                        stagingStatements.put(OpType.UPSERT, stmt);
+                    }
                 } else {
-                    stmt = prepareForAppendOnly(row);
-                }
-
-                try {
-                    LOG.debug("Executing statement: {}", stmt);
-                    stmt.executeUpdate();
-                    stmt.clearParameters();
-                } catch (SQLException e) {
-                    throw Status.INTERNAL
-                            .withDescription(
-                                    String.format(ERROR_REPORT_TEMPLATE, stmt, e.getMessage()))
-                            .asRuntimeException();
+                    var stmt = prepareInsertStatement(row);
+                    stagingStatements.put(OpType.INSERT, stmt);
                 }
             } catch (Exception e) {
                 throw new RuntimeException(e);
             }
         }
-    }
 
-    private PreparedStatement prepareForUpsert(SinkRow row) {
-        PreparedStatement stmt;
-        if (row.getOp() == Data.Op.DELETE) {
-            stmt = prepareDeleteStatement(row);
-        } else {
-            stmt = prepareUpsertStatement(row);
+        try {
+            // Execute staging statements after all rows are prepared.
+            // Note that we execute statement in the order of DELETE, UPSERT, INSERT as defined in
+            // OpType.
+            for (var entry : stagingStatements.entrySet()) {
```

Comment:
The order enforcement here is too implicit. Why not just use 3 statements variable for each op type and then execute them one by one in the required order.

Reply:
My main concern is that we need to track whether a statement has been filled with parameters before, otherwise we may execute a prepared statement without binding the placeholders, which I think may cause errors. So I resort to a map which act as an index for binded statements.

Reply:
If we instead use three local statement variables rather than a local tree map, it's the same. The variable is initially null, and will be set a value when any row binds on the statement. And then the three statements can be executed one by one in the required order when they are not null.

Reply:
You are right, but the code may become verbose as before, if check then execute for the 3 variables.

Reply:
If we want to ensure the execution order, I think it's better to keep the code verbose for better readability. Now the execution order is implicitly specified in an enum defined far away. And repeat the code of check and execute for only 3 times look no difference to me. Moreover, if we put the null check to the `executeStatement`, only the call to `executeStatement` will repeat 3 times, and no repeated null check at all.

Reply:
LGTM, will fix then